desc/derivatives.py

```diff
@@ -323,9 +326,13 @@ def _set_mode(self, mode) -> None:
 
         self._mode = mode
         if self._mode == "fwd":
-            self._compute = jax.jacfwd(self._fun, self._argnum)
+            self._compute = jacfwd_chunked(
```

Comment:
A question about this. So, basically, now we never use `jax.jacfwd` or `jax.jacrev`, instead calculate the jacobian from jvp or vjp's by `batched_vectorize` manor. Then, why do we have new functions `jacfwd_chunked` and `jacrev_chunked`? As far as I understand, they basically do `batched_vectorize` in another way. For `blocked` can't we just call each sub-objective `jvp_` method and get the whole thing? Also, I think the `jvp_` and `jac_` methods in our code kind of used in an interchangeable way, right? If we give `v` for each parameter like `jnp.eye(dim_x)` to `jvp_` then it is basically `jac_`. I point out this because it gets complicated once we start wrapping stuff. Like, `LinearConstraintProjection` uses `jvp_` of objective for `jac_` but not for its own `jvp_`. I see this PR gets rid of the difference between `blocked` and `batched` but I still think we need a proper dev guide for this. How do we calculate jacobians? For each type of `_Objective` and `ObjectiveFunction`...

Reply:
`jacfwd_chunked` and `jacrev_chunked` are still used for `_Objective.jac`. It's a bit confusing since under the hood, `jacfwd` is literally just doing `vmap(jvp(fun))`. It's also messy because in JAX parlance, jvps (aka A@x) are always computed with forward mode AD, while vjp (x@A = A.T@x) are always done with reverse mode.

This PR updates things so that the `jvp` methods of objectives will now sometimes use reverse mode when efficient or required (by getting the full jacobian with jacrev and then doing a matmul). 

Here's a rough outline of how everything works now:
- `_Objective`:
  - `_Objective.jac` method uses `jacfwd` or `jacrev` (chunked) so materializes the full matrix using either forward or reverse mode as appropriate
  - `_Objective.jvp` computes `A@x`. If the objective uses forward mode, then this basically just calls `jax.jvp` which uses forward mode. However if the objective uses reverse mode, we use `_Objective.jac` to get the full matrix and then matmul. This avoids trying to do forward mode on objectives for which that is innefficient or undefined, and vice versa.
- `ObjectiveFunction`
  - `ObjectiveFunction.jac` always just calls batched `ObjectiveFunction.jvp`, see below
  - `ObjectiveFunction.jvp` depends on whether `deriv_mode` is "batched" or "blocked". If "batched", then this basically just calls `jax.jvp` on `ObjectiveFunction.compute`, using forward mode for everything. If "blocked" then it calls `_Objective.jvp` for each objective, which as noted above will use forward or reverse mode as appropriate.
- `LinearConstraintProjection`, `ProximalProjection` etc.
   - These objectives are basically wrappers, so their jacobian looks like `A@B` where `A` is the jacobian of the underlying `ObjectiveFunction` and `B` is the transformation (projection etc). 
   - For `*Projection.jac`, we do this as a batched `ObjectiveFunction.jvp`, so `A@x` where `x=B`, and it uses whatever deriv mode the underlying `ObjectiveFunction` and `_Objective` wants (ie batched/blocked/fwd/rev)
   - `*Projection.jvp` also just passes through to `ObjectiveFunction.jvp` but modifies the tangent, like `A@(B@x) = A@y,    y=B@x`

Reply:
Thanks for the explanation. So, if the derivative we want is forward mode, then jvp or jac calls will result in a vectorized jvp (not always batched_vectorize though?). If it is reverse mode, than full jacobian is calculated in reverse mode ad and if vjp, the matmul is taken.

In a nutshell, can we say `jac_` method is a wrapper for jvp and vjps, such that if everthing was forward mode, we wouldn't need extra jac method?

Reply:
i mean we never really need the `jac` methods, since they can always be written as a vectorized jvp/vjp, but its annoying to have to create an identity matrix of the right type each time.

The vectorization/vmapping is now always batched/chunked